### PR TITLE
Fix issue 197

### LINKED ISSUE #197: Adding two int32 values results in int64, which is wrong [UDTF]
Consider the following example:
```
    @omnisci('int32(Column<int32>, RowMultiplier, OutputColumn<int32>)')
    def redefined_udtf2(x, m, y):
        for i in range(len(x)):
            y[i] = x[i] + np.int32(3)
        return len(x)
```
Running
```
    descr, result = omnisci.sql_execute(f'select * from table(redefined_udtf2(cursor(select i4 from {omnisci.table_name}), 1))')
```
fails with:
```
cannot store i64 to i32*: mismatching types
```

<details>

```
>       descr, result = omnisci.sql_execute(f'select * from table(redefined_udtf2(cursor(select i4 from {omnisci.table_name}), 1))')

rbc/tests/test_omnisci_column_basic.py:385: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
rbc/omniscidb.py:391: in sql_execute
    self.register()
rbc/omniscidb.py:708: in register
    OmnisciCompilerPipeline, self.debug)
rbc/irtools.py:271: in compile_to_LLVM
    pipeline_class=pipeline_class)
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler.py:551: in compile_extra
    return pipeline.compile_extra(func)
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler.py:331: in compile_extra
    return self._compile_bytecode()
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler.py:393: in _compile_bytecode
    return self._compile_core()
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler.py:373: in _compile_core
    raise e
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler.py:364: in _compile_core
    pm.run(self.state)
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler_machinery.py:347: in run
    raise patched_exception
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler_machinery.py:338: in run
    self._runPass(idx, pass_inst, state)
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler_lock.py:32: in _acquire_compile_lock
    return func(*args, **kwargs)
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler_machinery.py:302: in _runPass
    mutated |= check(pss.run_pass, internal_state)
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/compiler_machinery.py:275: in check
    mangled = func(compiler_state)
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/typed_passes.py:95: in run_pass
    raise_errors=self._raise_errors)
../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/typed_passes.py:67: in type_inference_stage
    infer.propagate(raise_errors=raise_errors)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <numba.typeinfer.TypeInferer object at 0x7fb13747e890>, raise_errors = True

    def propagate(self, raise_errors=True):
        newtoken = self.get_state_token()
        oldtoken = None
        # Since the number of types are finite, the typesets will eventually
        # stop growing.
    
        while newtoken != oldtoken:
            self.debug.propagate_started()
            oldtoken = newtoken
            # Errors can appear when the type set is incomplete; only
            # raise them when there is no progress anymore.
            errors = self.constraints.propagate(self)
            newtoken = self.get_state_token()
            self.debug.propagate_finished()
        if errors:
            if raise_errors:
                force_lit_args = [e for e in errors
                                  if isinstance(e, ForceLiteralArg)]
                if not force_lit_args:
>                   raise errors[0]
E                   numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
E                   Invalid use of Function(<built-in function setitem>) with argument(s) of type(s): (STRUCT__iPLK, int64, int64)
E                    * parameterized
E                   In definition 0:
E                       All templates rejected with literals.
E                   In definition 1:
E                       All templates rejected without literals.
E                   In definition 2:
E                       All templates rejected with literals.
E                   In definition 3:
E                       All templates rejected without literals.
E                   In definition 4:
E                       All templates rejected with literals.
E                   In definition 5:
E                       All templates rejected without literals.
E                   In definition 6:
E                       LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
E                   cannot store i64 to i32*: mismatching types
E                   
E                   File "rbc/omnisci_backend/omnisci_buffer.py", line 219:
E                   def omnisci_buffer_setitem(a, i, v):
E                       <source elided>
E                       if isinstance(a, BufferType):
E                           return lambda a, i, v: omnisci_buffer_setitem_(a, i, v)
E                           ^
E                   
E                   [1] During: lowering "$10call_function.4 = call $2load_global.0(a, i, v, func=$2load_global.0, args=[Var(a, omnisci_buffer.py:219), Var(i, omnisci_buffer.py:219), Var(v, omnisci_buffer.py:219)], kws=(), vararg=None)" at /home/pearu/git/xnd-project/rbc/rbc/omnisci_backend/omnisci_buffer.py (219)
E                       raised from /home/pearu/miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/six.py:669
E                   In definition 7:
E                       LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
E                   cannot store i64 to i32*: mismatching types
E                   
E                   File "rbc/omnisci_backend/omnisci_buffer.py", line 219:
E                   def omnisci_buffer_setitem(a, i, v):
E                       <source elided>
E                       if isinstance(a, BufferType):
E                           return lambda a, i, v: omnisci_buffer_setitem_(a, i, v)
E                           ^
E                   
E                   [1] During: lowering "$10call_function.4 = call $2load_global.0(a, i, v, func=$2load_global.0, args=[Var(a, omnisci_buffer.py:219), Var(i, omnisci_buffer.py:219), Var(v, omnisci_buffer.py:219)], kws=(), vararg=None)" at /home/pearu/git/xnd-project/rbc/rbc/omnisci_backend/omnisci_buffer.py (219)
E                       raised from /home/pearu/miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/six.py:669
E                   In definition 8:
E                       All templates rejected with literals.
E                   In definition 9:
E                       All templates rejected without literals.
E                   In definition 10:
E                       All templates rejected with literals.
E                   In definition 11:
E                       All templates rejected without literals.
E                   This error is usually caused by passing an argument of a type that is unsupported by the named function.
E                   [1] During: typing of setitem at /home/pearu/git/xnd-project/rbc/rbc/tests/test_omnisci_column_basic.py (382)
E                   
E                   File "rbc/tests/test_omnisci_column_basic.py", line 382:
E                       def redefined_udtf2(x, m, y):
E                           <source elided>
E                           for i in range(len(x)):
E                               y[i] = x[i] + np.int32(3)
E                               ^

../../../miniconda3/envs/rbc-dev/lib/python3.7/site-packages/numba/typeinfer.py:985: TypingError
```

</details>

The following works:
```
    @omnisci('int32(Column<int32>, RowMultiplier, OutputColumn<int32>)')
    def redefined_udtf2(x, m, y):
        for i in range(len(x)):
            y[i] = np.int32(x[i] + 3)
        return len(x)
```

### PR DESCRIPTION
**Pull request by guilhermeleobas**
As title. Fix #197

### CODE REVIEW

File: rbc/omnisci_backend/omnisci_buffer.py [hunk 1]
```diff
@@ -177,6 +177,29 @@ def omnisci_buffer_getitem(x, i):
...
+
+        if isinstance(nb_value, types.Integer):
+            if eltype.bitwidth < nb_value.bitwidth:
+                return builder.trunc(value, buf_typ)
+            elif eltype.bitwidth > nb_value.bitwidth:
+                is_signed = nb_value.signed
+                return builder.zext(value, buf_typ) if is_signed else \
+                    builder.sext(value, buf_typ)
+            return value
+        else:  # Floating-point
```
Comment (pearu) [inline rbc/omnisci_backend/omnisci_buffer.py:196]: Could you add more tests to cover all branches of the truncate_or_extend? `int32[pos] = int32 + int64`, `int64[pos] = int32 + int64`, `int16[pos] = ...`, etc. Same for floats.

Also, is this code complete when `nb_value` is `bool`, or for the future, `str` or `array`?

**Review by pearu — CHANGES_REQUESTED** [review]
LGTM, but add more tests.

File: rbc/tests/test_omnisci_array.py [hunk 2]
```diff
@@ -409,3 +409,61 @@ def array_is_null(size):
...
+        for i in range(len(x)):
+            y[i] = types.float64(x[i] + 3.0)  # fptrunc(f64) -> f3232
+        return y
+
+    @omnisci('double[](double[])')
+    def fn_issue197_float_fpext(x):
+        y = np.zeros_like(x)
+        for i in range(len(x)):
+            y[i] = types.float32(x[i] + 3.0)  # fpext(f32) -> f64
+        return y
```
Comment (pearu) [inline rbc/tests/test_omnisci_array.py:460]: You can join all these functions into one to reduce the compilation burden and code:
```
cast = dict(trunc=types.int64, sext=types.int8, zext=types.uint8, fptrunc=types.float64, fpext=types.float32)[suffix]

@omnisci(f'{typ}[]({typ}[])')
def fn_issue197(x):
    y = np.zeros_like(x)
    for i in range(len(x)):
        y[i] = cast(x[i] + 3)
    return y

fn_name = f"fn_issue197_{typ}_{suffix}"
fn_issue197.__name__ = fn_name

```

**Review by pearu — CHANGES_REQUESTED** [review]
Looks good. Thanks for the tests!
I have one suggestion to reduce the test code. Also, could you test the original problem of the issue #197, that was a UDTF definition:
```
    @omnisci('int32(Column<int32>, RowMultiplier, OutputColumn<int32>)')
    def redefined_udtf2(x, m, y):
        for i in range(len(x)):
            y[i] = x[i] + 3
        return len(x)
```
(notice no type casting in `y[i] = x[i] + 3`).

File: rbc/omnisci_backend/omnisci_buffer.py [hunk 3]
```diff
@@ -195,6 +225,9 @@ def codegen(context, builder, signature, args):
 def omnisci_buffer_setitem_(typingctx, data, index, value):
     sig = types.none(data, index, value)
 
+    eltype = data.members[0].dtype
+    nb_value = value
```
Comment (guilhermeleobas) [inline rbc/omnisci_backend/omnisci_buffer.py:229]: I'd like to use `data.eltype` here but `BufferType` do not hold this information

**Review by guilhermeleobas — COMMENTED** [review]
Hi Pearu, I've added the original test case as well tests for `bool` values.

File: rbc/omnisci_backend/omnisci_buffer.py [hunk 4]
```diff
@@ -173,10 +180,39 @@ def omnisci_buffer_getitem(x, i):
...
+    elif isinstance(nb_value, types.Float):  # Floating-point
+        if eltype.bitwidth < nb_value.bitwidth:
+            return builder.fptrunc(value, buf_typ)  # truncate
+        elif eltype.bitwidth > nb_value.bitwidth:
+            return builder.fpext(value, buf_typ)  # extend
+    elif isinstance(nb_value, types.Boolean):
+        if buf_typ.width < value.type.width:
+            return builder.trunc(value, buf_typ)
+        elif buf_typ.width > value.type.width:
+            return builder.sext(value, buf_typ)
```
Comment (pearu) [inline rbc/omnisci_backend/omnisci_buffer.py:204]: I wonder in which circumstances are these returns effective?

**Review by pearu — APPROVED** [review]
LGTM

File: rbc/omnisci_backend/omnisci_buffer.py (hunk 4 above)
Comment (guilhermeleobas) [reply rbc/omnisci_backend/omnisci_buffer.py:204]:
```python
@omnisci('bool[](bool[])')
def fn_issue197_bool(x):
    y = np.zeros_like(x)
    for i in range(len(x)):
        y[i] = bool(x[i])
    return y
```

In the test above, `bool` casts the value to `int1`.

**Pull request merged**